src/Core/Form/IdentifiableObject/CommandBuilder/Product/UpdateProductCommandsBuilder.php

```diff
@@ -0,0 +1,177 @@
+use PrestaShop\PrestaShop\Core\Domain\Product\Command\UpdateProductInput\ProductDetailsInput;
+use PrestaShop\PrestaShop\Core\Domain\Product\Command\UpdateProductInput\ProductOptionsInput;
+use PrestaShop\PrestaShop\Core\Domain\Product\Command\UpdateProductInput\ProductPricesInput;
+use PrestaShop\PrestaShop\Core\Domain\Product\ValueObject\ProductId;
+use PrestaShop\PrestaShop\Core\Domain\Shop\ValueObject\ShopConstraint;
+use PrestaShop\PrestaShop\Core\Form\IdentifiableObject\CommandBuilder\CommandBuilder;
+use PrestaShop\PrestaShop\Core\Form\IdentifiableObject\CommandBuilder\CommandBuilderConfig;
+use PrestaShop\PrestaShop\Core\Form\IdentifiableObject\CommandBuilder\DataField;
+
+//@todo: abit confusing when it should always build only one command but is called plural and returns array. smth needs adjustments
+class UpdateProductCommandsBuilder implements MultiShopProductCommandsBuilderInterface
+{
+    /**
+     * {@inheritDoc}
+     */
+    public function buildCommands(ProductId $productId, array $formData, ShopConstraint $shopConstraint): array
+    {
+        $updateProductCommand = new UpdateProductCommand($productId->getValue(), $shopConstraint);
+
+        $updateProductCommand
+            ->setBasicInformation($this->buildBasicInfo($formData))
+            ->setOptions($this->buildOptions($formData))
+            ->setPrices($this->buildPrices($formData))
+            ->setDetails($this->buildDetails($formData))
+        ;
+
+        return [$updateProductCommand];
+    }
+
+    /**
+     * @param array<string, mixed> $formData
+     *
+     * @return BasicInformationInput|null
+     */
+    private function buildBasicInfo(array $formData): ?BasicInformationInput
+    {
+        if (empty($formData['description']) && empty($formData['header']['name'])) {
+            return null;
+        }
+
+        $config = new CommandBuilderConfig();
+        $config
+            ->addMultiShopField('[header][name]', 'setLocalizedNames', DataField::TYPE_ARRAY)
+            ->addMultiShopField('[description][description]', 'setLocalizedDescriptions', DataField::TYPE_ARRAY)
+            ->addMultiShopField('[description][description_short]', 'setLocalizedShortDescriptions', DataField::TYPE_ARRAY)
+        ;
+
+        $commandBuilder = new CommandBuilder($config);
+        $input = new BasicInformationInput();
+        $inputs = $commandBuilder->buildCommands($formData, $input);
+
+        return $inputs[0] ?? null;
+    }
+
+    /**
+     * @param array<string, mixed> $formData
+     *
+     * @return ProductOptionsInput|null
+     */
+    private function buildOptions(array $formData): ?ProductOptionsInput
+    {
+        if (empty($formData['options']) &&
+            !isset($formData['description']['manufacturer']) &&
+            !isset($formData['specifications'])) {
+            return null;
+        }
+
+        $config = new CommandBuilderConfig();
```

Comment:
this helps us to hide the repetitive checks of isset and the type casting (and probably some other stuff). By the naming - it suppose to build the commands, but actually works for building any DTO's, should we care about the naming? Haven't yet checked how it will work with multishop (as now the shop constraint should only be relevant in the UpdateProductHandler, and not anymore in these DTO's)